In [1]:
import os
import numpy as np
import pandas as pd
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from sklearn.ensemble import RandomForestRegressor, VotingRegressor
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler


os.chdir('C:/Users/kksp1/OneDrive/Desktop/3_2학기 해야할 일/수업/3-2/머신러닝/미니프로젝트/풀젝_4')
#print(os.getcwd())
input_file = "winequality-red.csv"
wine = pd.read_csv(input_file)

In [2]:
wine.head(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

## 상관계수 확인

In [3]:
corr_matrix = wine.corr()
corr_matrix["quality"].sort_values(ascending=False)

quality                 1.000000
alcohol                 0.476166
sulphates               0.251397
citric acid             0.226373
fixed acidity           0.124052
residual sugar          0.013732
free sulfur dioxide    -0.050656
pH                     -0.057731
chlorides              -0.128907
density                -0.174919
total sulfur dioxide   -0.185100
volatile acidity       -0.390558
Name: quality, dtype: float64

## 특성 추가

In [4]:
wine["balance"] = wine["alcohol"] - wine["sulphates"] - wine["volatile acidity"]
wine['total_minerals'] = wine['chlorides'] + wine['sulphates']
wine['sulph_sub_chlor']=wine['sulphates']-wine['chlorides']

### drop & 분리

In [5]:
temp = wine.copy().drop("residual sugar", axis=1)
wine = temp.drop("quality", axis=1)
wine_labels = temp["quality"].copy()

## Scaling

In [6]:
sc = StandardScaler()
wine = sc.fit_transform(wine)

In [7]:
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())
    print("Standard deviation:", scores.std())

## 모델 학습

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error

# 탐색할 하이퍼파라미터 그리드 정의
param_grid = {
    'n_estimators': [500, 1000, 1500],
    'max_depth': [20, 30, 40],
    'max_features': [3, 5, 7],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# ExtraTreesRegressor 모델 초기화
extra_trees = ExtraTreesRegressor(random_state=42)

# GridSearchCV로 최적의 하이퍼파라미터 탐색
grid_search = GridSearchCV(extra_trees, param_grid, cv=5, scoring='neg_mean_squared_error', n_jobs=-1)
grid_search.fit(sh_t_data, sh_t_target)

# 최적의 하이퍼파라미터 출력
print("Best parameters found:", grid_search.best_params_)

# 최적의 모델 저장
best_model = grid_search.best_estimator_

# 최적의 모델을 사용한 예측 및 평가
predictions = best_model.predict(sh_t_data)
mse = mean_squared_error(sh_t_target, predictions)
rmse = np.sqrt(mse)
print("RMSE of the best model:", rmse)

#Best parameters found: {'max_depth': 30, 'max_features': 3, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 1500}
#RMSE of the best model: 0.0009072077184445359

In [9]:
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import cross_val_score

seed = 42
np.random.seed(seed)
sh_in = np.random.permutation(len(wine))
sh_t_data = wine[sh_in]
sh_t_target = wine_labels.values[sh_in]

extra_trees_clf = ExtraTreesRegressor(max_depth = 30, max_features = 3, min_samples_leaf = 1, min_samples_split = 2,
                                      n_estimators =1500, random_state=42)

scores = cross_val_score(extra_trees_clf, sh_t_data, sh_t_target, scoring="neg_mean_squared_error", cv=10, n_jobs=-1)
rmse_scores = np.sqrt(-scores)
display_scores(rmse_scores)

Scores: [0.54179384 0.52987146 0.57293101 0.54434763 0.55461333 0.58123847
 0.52684393 0.58453109 0.505835   0.5163231 ]
Mean: 0.5458328866741197
Standard deviation: 0.025820604789053292


In [47]:
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import cross_val_score

seed = 42
np.random.seed(seed)
sh_in = np.random.permutation(len(wine))
sh_t_data = wine[sh_in]
sh_t_target = wine_labels.values[sh_in]

extra_trees_clf = ExtraTreesRegressor(max_depth = 31, max_features = 3, min_samples_leaf = 1, min_samples_split = 2,
                                      n_estimators =1580, random_state=42)

scores = cross_val_score(extra_trees_clf, sh_t_data, sh_t_target, scoring="neg_mean_squared_error", cv=10, n_jobs=-1)
rmse_scores = np.sqrt(-scores)
display_scores(rmse_scores)

Scores: [0.5418525  0.53056773 0.57265176 0.54322405 0.55372366 0.58103577
 0.52628959 0.58379052 0.50574029 0.51628641]
Mean: 0.5455162259496422
Standard deviation: 0.02564871162633668


#### n_estimators
- 1500 Mean: 0.5458328866741197
- 1600 Mean: 0.545782189760157
- 1700 Mean: 0.545995864745351
- 1550 Mean: 0.5457929352123588
- 1570 Mean: 0.5457845125204035
- 1580 Mean: 0.5457390596646761
- 1590 Mean: 0.5457455310332243
- 1585 Mean: 0.5457421432600117
- 1581 Mean: 0.5457442316855113

#### max_depth
- 30 Mean: 0.5457390596646761
- 31 Mean: 0.5455162259496422
- 32 Mean: 0.5456229411538029
- 29 Mean: 0.5457076320124534

#### max_features
- 3 Mean: 0.5455162259496422
- 4 Mean: 0.5469516095796015
- 2 Mean: 0.5469751347904263

#### min_samples_leaf
- 1 Mean: 0.5455162259496422
- 3 Mean: 0.5910591480459194

### 시드 찾기

In [48]:
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import cross_val_score
import numpy as np

best_seed = None
best_mean_rmse = np.inf

# 100번의 시드를 적용하여 교차 검증 실행
for i in range(0, 2880 * 123, 123):
    seed = 42 + i  # 시드 값 설정
    np.random.seed(seed)
    sh_in = np.random.permutation(len(wine))
    sh_t_data = wine[sh_in]
    sh_t_target = wine_labels.values[sh_in]

    extra_trees_clf = ExtraTreesRegressor(
        max_depth=31, max_features=3, min_samples_leaf=1, min_samples_split=2,
        n_estimators=1580, random_state=42
    )

    scores = cross_val_score(
        extra_trees_clf, sh_t_data, sh_t_target, scoring="neg_mean_squared_error", cv=10, n_jobs=-1
    )
    rmse_scores = np.sqrt(-scores)
    mean_rmse = np.mean(rmse_scores)

    print(f"Seed: {seed}, Mean RMSE: {mean_rmse}")

    if mean_rmse < best_mean_rmse:
        best_mean_rmse = mean_rmse
        best_seed = seed

print(f"Best Seed: {best_seed}, Best Mean RMSE: {best_mean_rmse}")


Seed: 42, Mean RMSE: 0.5455162259496422
Seed: 165, Mean RMSE: 0.5548472013586865
Seed: 288, Mean RMSE: 0.5534616797071397
Seed: 411, Mean RMSE: 0.5543616727498633
Seed: 534, Mean RMSE: 0.5560626313071555
Seed: 657, Mean RMSE: 0.5455578776877339
Seed: 780, Mean RMSE: 0.5492069472787613
Seed: 903, Mean RMSE: 0.553920352425378
Seed: 1026, Mean RMSE: 0.5467309046422317
Seed: 1149, Mean RMSE: 0.5508437576507987
Seed: 1272, Mean RMSE: 0.547936887394394
Seed: 1395, Mean RMSE: 0.5458151805741913
Seed: 1518, Mean RMSE: 0.5461771942341709
Seed: 1641, Mean RMSE: 0.5471054297717256
Seed: 1764, Mean RMSE: 0.549907201517312
Seed: 1887, Mean RMSE: 0.552779111614224
Seed: 2010, Mean RMSE: 0.5513085839794232
Seed: 2133, Mean RMSE: 0.5486650919733057
Seed: 2256, Mean RMSE: 0.5474807768317592
Seed: 2379, Mean RMSE: 0.5502765949489626
Seed: 2502, Mean RMSE: 0.5460449931900986
Seed: 2625, Mean RMSE: 0.5455774299303394
Seed: 2748, Mean RMSE: 0.5487188193629245
Seed: 2871, Mean RMSE: 0.550551035629065
Seed: 

In [49]:

from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import cross_val_score

seed = 231651
np.random.seed(seed)
sh_in = np.random.permutation(len(wine))
sh_t_data = wine[sh_in]
sh_t_target = wine_labels.values[sh_in]

extra_trees_clf = ExtraTreesRegressor(max_depth = 30, max_features = 3, min_samples_leaf = 1, min_samples_split = 2,
                                      n_estimators =1500, random_state=42)

scores = cross_val_score(extra_trees_clf, sh_t_data, sh_t_target, scoring="neg_mean_squared_error", cv=10, n_jobs=-1)
rmse_scores = np.sqrt(-scores)
display_scores(rmse_scores)

Scores: [0.54123482 0.58457523 0.53867009 0.52293964 0.40888372 0.59492893
 0.50508179 0.53689945 0.53388685 0.61126049]
Mean: 0.5378361009928498
Standard deviation: 0.05365395249765184


### XGBRegressor 모델로 교체

In [64]:
from xgboost import XGBRegressor

seed = 231651
np.random.seed(seed)
sh_in = np.random.permutation(len(wine))
sh_t_data = wine[sh_in]
sh_t_target = wine_labels.values[sh_in]

# Define XGBoost Regressor
xgboost_clf = XGBRegressor(max_depth=30, n_estimators=1500, random_state=42)

scores = cross_val_score(xgboost_clf, sh_t_data, sh_t_target, scoring="neg_mean_squared_error", cv=10, n_jobs=-1)
rmse_scores = np.sqrt(-scores)
display_scores(rmse_scores)


Scores: [0.60476494 0.65639111 0.5917852  0.57768131 0.4848211  0.63435674
 0.57650795 0.64378112 0.5718662  0.65747491]
Mean: 0.5999430593779425
Standard deviation: 0.049777845893586845


In [65]:
from xgboost import XGBRegressor
from sklearn.model_selection import GridSearchCV

# Define the parameter grid
param_grid = {
    'max_depth': [10, 20, 30],
    'n_estimators': [1000, 1500, 2000],
    'learning_rate': [0.1, 0.01, 0.001],
    'min_child_weight': [1, 3, 5]
}

# Create XGBRegressor
xgboost_clf = XGBRegressor(random_state=42)

# Perform GridSearchCV
grid_search = GridSearchCV(estimator=xgboost_clf, param_grid=param_grid, 
                           scoring='neg_mean_squared_error', cv=10, n_jobs=-1)
grid_search.fit(sh_t_data, sh_t_target)

# Get the best parameters and best score
best_params = grid_search.best_params_
best_score = np.sqrt(-grid_search.best_score_)

print("Best Parameters:", best_params)
print("Best RMSE:", best_score)


Best Parameters: {'learning_rate': 0.01, 'max_depth': 10, 'min_child_weight': 5, 'n_estimators': 1500}
Best RMSE: 0.5773478507504012


### 최종 선택

In [9]:
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import cross_val_score

seed = 553419
np.random.seed(seed)
sh_in = np.random.permutation(len(wine))
sh_t_data = wine[sh_in]
sh_t_target = wine_labels.values[sh_in]

extra_trees_clf = ExtraTreesRegressor(max_depth = 30, max_features = 3, min_samples_leaf = 1, min_samples_split = 2,
                                      n_estimators =1500, random_state=42)

scores = cross_val_score(extra_trees_clf, sh_t_data, sh_t_target, scoring="neg_mean_squared_error", cv=10, n_jobs=-1)
rmse_scores = np.sqrt(-scores)
display_scores(rmse_scores)

Scores: [0.534946   0.59027431 0.51421488 0.5202973  0.57018372 0.63542438
 0.46442906 0.5036699  0.5515057  0.49015569]
Mean: 0.537510094302253
Standard deviation: 0.04817537868057476


In [18]:
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import cross_val_score

seed = 553419
np.random.seed(seed)
sh_in = np.random.permutation(len(wine))
sh_t_data = wine[sh_in]
sh_t_target = wine_labels.values[sh_in]

extra_trees_clf = ExtraTreesRegressor(max_depth = 26, max_features = 4, min_samples_leaf = 1, min_samples_split = 2,
                                      n_estimators =810, random_state=91795)

scores = cross_val_score(extra_trees_clf, sh_t_data, sh_t_target, scoring="neg_mean_squared_error", cv=10, n_jobs=-1)
rmse_scores = np.sqrt(-scores)
display_scores(rmse_scores)

Scores: [0.52913696 0.59249136 0.51494087 0.5207042  0.5665532  0.6318132
 0.46590324 0.50700265 0.54229186 0.48888002]
Mean: 0.5359717550979181
Standard deviation: 0.04695078951015774


## Mean: 0.5359717550979181